File: Problems/052-n-queens-ii/n-queens-ii.py
class Solution:
    def __init__(self):
            self.count = 0

    def totalNQueens(self, n: int) -> int:
        def backtrack(r=0):
            if r == n:
                self.count += 1
                return

            for c in range(n):
                id1 = c - r + n
                id2 = 2 * n - c - r - 1
                if not col[c] and not d1[id1] and not d2[id2]:
                    col[c], d1[id1], d2[id2] = True, True, True
                    backtrack(r + 1)
                    col[c], d1[id1], d2[id2] = False, False, False

        col = [False] * n
        d1 = [False] * 2 * n
        d2 = [False] * 2 * n
        backtrack()

        return self.count


print(Solution().totalNQueens(4))

File: Problems/052-n-queens-ii/n-queens-ii.go
package problem052

func totalNQueens(n int) int {
	col := make([]bool, n)
	d1 := make([]bool, 2*n)
	d2 := make([]bool, 2*n)
	res := 0
	backtrack(&res, 0, n, col, d1, d2)

	return res
}

func backtrack(res *int, r, n int, col, d1, d2 []bool) {
	if r == n {
		*res++
		return
	}

	for c := 0; c < n; c++ {
		id1 := r - c + n
		id2 := 2*n - r - c - 1
		if !col[c] && !d1[id1] && !d2[id2] {
			col[c], d1[id1], d2[id2] = true, true, true
			backtrack(res, r+1, n, col, d1, d2)
			col[c], d1[id1], d2[id2] = false, false, false
		}
	}
}
